pkg/controller/sidecarset/sidecarset_utils.go

```diff
@@ -101,3 +109,212 @@ func inconsistentStatus(sidecarSet *appsv1alpha1.SidecarSet, status *appsv1alpha
 		status.UpdatedPods != sidecarSet.Status.UpdatedPods ||
 		status.ReadyPods != sidecarSet.Status.ReadyPods
 }
+
+// add this cache to avoid be influenced by informer cache latency when controller try to count maxUnavailable
+type updatedPodCache struct {
+	lock sync.RWMutex
+	// key is in the format: sidecarset name/pod namespace/pod name
+	// value is sidecarset hash
+	podSidecarUpdated map[string]string
+}
+
+func (u *updatedPodCache) set(key, hash string) {
+	u.lock.Lock()
+	u.podSidecarUpdated[key] = hash
+	u.lock.Unlock()
+}
+
+func (u *updatedPodCache) delete(key string) {
+	u.lock.Lock()
+	delete(u.podSidecarUpdated, key)
+	u.lock.Unlock()
+}
+
+func (u *updatedPodCache) reset() {
+	u.lock.Lock()
+	u.podSidecarUpdated = make(map[string]string)
+	u.lock.Unlock()
+}
+
+// remove cache related with sidecarset/pod does not exist anymore
+func (u *updatedPodCache) removeStaleCache(sidecarSet *appsv1alpha1.SidecarSet, pods []*corev1.Pod) {
+	keyIsValid := make(map[string]bool, len(pods))
+	for _, pod := range pods {
+		key := fmt.Sprintf("%v/%v/%v", sidecarSet.Name, pod.Namespace, pod.Name)
+		keyIsValid[key] = true
+	}
+
+	u.lock.Lock()
+	for key := range u.podSidecarUpdated {
+		if !keyIsValid[key] {
+			delete(u.podSidecarUpdated, key)
+		}
+	}
+	u.lock.Unlock()
+}
+
+// return true means: sidecar of pod is updated
+func (u *updatedPodCache) isSidecarUpdated(key, curHash string) bool {
+	u.lock.RLock()
+	hash := u.podSidecarUpdated[key]
+	u.lock.RUnlock()
+
+	if hash == curHash {
+		return true
+	}
+	// if sidecarset changed, clean all stale cache
+	if hash != "" && hash != curHash {
```

Comment:
no need for hash!=curHash. there is a shortcut at line 163 for the equal case.

Reply:
it's true, but it will be easy to understand

Reply:
done